File: Hawkeye-Python code/bitmask.py
import math

def bitmask(l):
    return (1 << l) - 1 if l < 64 else 0xFFFFFFFFFFFFFFFF

def bits(x, i, l):
    return (x >> i) & bitmask(l)

def SAMPLED_SET(set_num):
    return bits(set_num, 0, 6) == bits(set_num, int(math.log2(4096)) - 6, 6)

# Test the bitmask and bits function
for l in range(1, 65):
    print(f"bitmask({l}): {bitmask(l)}")

test_val = 0xF0F0F0F0F0F0F0F0  # Sample value for testing
print(f"bits({test_val}, 4, 8): {bits(test_val, 4, 8)}")

# Test the SAMPLED_SET function
for set_num in range(100):
    print(f"Set {set_num} is {'sampled' if SAMPLED_SET(set_num) else 'not sampled'}")

File: Hawkeye-CPP-test codes/bitmask.cpp
#include <iostream>
#include <cmath>

#define bitmask(l) (((l) == 64) ? (unsigned long long)(-1LL) : ((1LL << (l))-1LL))
#define bits(x, i, l) (((x) >> (i)) & bitmask(l))
#define SAMPLED_SET(set) (bits(set, 0 , 6) == bits(set, ((unsigned long long)log2(4096) - 6), 6))

int main() {
    // Test the bitmask and bits function
    for (int l = 1; l <= 64; ++l) {
        std::cout << "bitmask(" << l << "): " << bitmask(l) << std::endl;
    }

    unsigned long long test_val = 0xF0F0F0F0F0F0F0F0;  // Sample value for testing
    std::cout << "bits(" << test_val << ", 4, 8): " << bits(test_val, 4, 8) << std::endl;

    // Test the SAMPLED_SET macro
    for (int set = 0; set < 100; ++set) {
        std::cout << "Set " << set << " is " << (SAMPLED_SET(set) ? "" : "not ") << "sampled." << std::endl;
    }

    return 0;
}
